src/dwarf.rs

```diff
@@ -957,6 +1007,123 @@ impl DwarfResolver {
     }
 
+    /// Extract the symbol information from DWARf if having not did it
+    /// before.
+    fn ensure_debug_info_syms(&self) -> Result<(), Error> {
+        if self.enable_debug_info_syms {
+            let mut dis_ref = self.debug_info_syms.borrow_mut();
+            if dis_ref.is_some() {
+                return Ok(());
+            }
+            let mut debug_info_syms = debug_info_parse_symbols(&self.parser, None, 1)?;
+            debug_info_syms.sort_by_key(|v: &DWSymInfo| -> &str { v.name });
+            *dis_ref = Some(unsafe { mem::transmute(debug_info_syms) });
+        }
+        Ok(())
+    }
+
+    /// Find the address of a symbol from DWARF.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - is the symbol name to find.
+    /// * `opts` - is the context giving additional parameters.
+    pub fn find_address(&self, name: &str, opts: &FindAddrOpts) -> Result<Vec<SymbolInfo>, Error> {
+        if let SymbolType::Variable = opts.sym_type {
+            return Err(Error::new(ErrorKind::Unsupported, "Not implemented"));
+        }
+        let r = self.parser.find_address(name, opts);
+        if r.is_ok() {
+            return r;
+        }
+
+        self.ensure_debug_info_syms()?;
+        let dis_ref = self.debug_info_syms.borrow();
+        let debug_info_syms = dis_ref.as_ref().unwrap();
+        let mut idx =
+            match debug_info_syms.binary_search_by_key(&name.to_string(), |v| v.name.to_string()) {
+                Ok(idx) => idx,
+                _ => {
+                    return Ok(vec![]);
+                }
+            };
+        while idx > 0 && debug_info_syms[idx].name.eq(name) {
+            idx -= 1;
+        }
+        if !debug_info_syms[idx].name.eq(name) {
+            idx += 1;
+        }
+        let mut found = vec![];
+        while debug_info_syms[idx].name.eq(name) {
+            let DWSymInfo {
+                address,
+                size,
+                sym_type,
+                ..
+            } = debug_info_syms[idx];
+            found.push(SymbolInfo {
+                name: name.to_string(),
+                address,
+                size,
+                sym_type,
+                ..Default::default()
+            });
+            idx += 1;
+        }
+        Ok(found)
+    }
+
+    /// Find the address of symbols matching a pattern from DWARF.
+    ///
+    /// #Arguments
+    ///
+    /// * `pattern` - is a regex pattern to match symbols.
+    /// * `opts` - is the context giving additional parameters.
+    ///
+    /// Return a list of symbols including addresses and other information.
+    pub fn find_address_regex(
+        &self,
+        pattern: &str,
+        opts: &FindAddrOpts,
+    ) -> Result<Vec<SymbolInfo>, Error> {
+        if let SymbolType::Variable = opts.sym_type {
+            return Err(Error::new(ErrorKind::Unsupported, "Not implemented"));
+        }
+        let r = self.parser.find_address_regex(pattern, opts)?;
+        if r.len() > 0 {
```

Comment:
is it correct to stop searching for symbols in DWARF if we found some symbols in ELF? I'd expect that if opts say we need to search DWARF, we search DWARF regardless if ELF has any matches or not? What do I miss?